diff --git a/arch/arm64/configs/ksu_manual.config b/arch/arm64/configs/ksu_manual.config
new file mode 100644
--- /dev/null
+++ b/arch/arm64/configs/ksu_manual.config
@@ -0,0 +1,2 @@
+CONFIG_KSU_MANUAL_HOOK=y
+# CONFIG_KSU_KPROBES_HOOK is not set
diff --git a/build.sh b/build.sh
--- a/build.sh
+++ b/build.sh
@@ -164,5 +164,5 @@ echo "Generating configuration file..."
 echo "-----------------------------------------------"
-make ${MAKE_ARGS} -j$CORES exynos9830_defconfig $MODEL.config $KSU $RECOVERY || abort
+make ${MAKE_ARGS} -j$CORES exynos9830_defconfig $MODEL.config $KSU ksu_manual.config $RECOVERY || abort
 
 if [ ! -z "$DTBS" ]; then
     MAKE_ARGS="$MAKE_ARGS dtbs"
diff --git a/fs/devpts/inode.c b/fs/devpts/inode.c
--- a/fs/devpts/inode.c
+++ b/fs/devpts/inode.c
@@ -599,10 +599,6 @@ struct dentry *devpts_pty_new(struct pts_fs_info *fsi, int index, void *priv)
 	return dentry;
 }
 
-#ifdef CONFIG_KSU
-extern int ksu_handle_devpts(struct inode*);
-#endif
-
 /**
  * devpts_get_priv -- get private data for a slave
  * @pts_inode: inode of the slave
@@ -611,10 +607,6 @@ extern int ksu_handle_devpts(struct inode*);
  */
 void *devpts_get_priv(struct dentry *dentry)
 {
-#ifdef CONFIG_KSU
-	ksu_handle_devpts(dentry->d_inode);
-#endif
-
 	if (dentry->d_sb->s_magic != DEVPTS_SUPER_MAGIC)
 		return NULL;
 	return dentry->d_fsdata;
diff --git a/kernel/reboot.c b/kernel/reboot.c
--- a/kernel/reboot.c
+++ b/kernel/reboot.c
@@ -304,6 +304,11 @@ EXPORT_SYMBOL_GPL(kernel_power_off);
 
 DEFINE_MUTEX(system_transition_mutex);
 
+#ifdef CONFIG_KSU
+extern int ksu_handle_sys_reboot(int magic1, int magic2, unsigned int cmd,
+				 void __user **arg);
+#endif
+
 /*
  * Reboot system call: for obvious reasons only root may call it,
  * and even root needs to set up some magic numbers in the registers
@@ -319,6 +324,10 @@ SYSCALL_DEFINE4(reboot, int, magic1, int, magic2, unsigned int, cmd,
 	char buffer[256];
 	int ret = 0;
 
+#ifdef CONFIG_KSU
+	ksu_handle_sys_reboot(magic1, magic2, cmd, &arg);
+#endif
+
 	/* We only trust the superuser with rebooting the system. */
 	if (!ns_capable(pid_ns->user_ns, CAP_SYS_BOOT))
 		return -EPERM;
